File: src/diagnostics.cpp
// -*-c++-*--------------------------------------------------------------------
// Copyright 2024 Bernd Pfrommer <bernd.pfrommer@gmail.com>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <cassert>
#include <fstream>
#include <iomanip>
#include <multicam_imu_calib/diagnostics.hpp>
#include <multicam_imu_calib/utilities.hpp>
#include <opencv2/calib3d.hpp>

namespace multicam_imu_calib
{
namespace diagnostics
{
std::vector<std::vector<std::array<double, 2>>> computeAllProjectedPoints(
  const std::vector<std::vector<std::array<double, 3>>> & world_pts,
  const std::vector<gtsam::Pose3> cam_world_poses,
  const Intrinsics & intrinsics, DistortionModel distortion_model,
  const DistortionCoefficients & distortion_coefficients)
{
  assert(cam_world_poses.size() == world_pts.size());

  std::vector<std::vector<std::array<double, 2>>> all_pts;
  for (size_t i = 0; i < cam_world_poses.size(); i++) {
    const auto & T_w_c = cam_world_poses[i];
    const auto ic = utilities::makeProjectedPoints(
      intrinsics, distortion_model, distortion_coefficients, T_w_c,
      world_pts[i]);
    all_pts.push_back(ic);
  }
  return (all_pts);
}

std::tuple<double, double, size_t> computeProjectionError(
  size_t cam_idx, const std::vector<uint64_t> & times,
  const std::vector<std::vector<std::array<double, 3>>> & world_pts,
  const std::vector<std::vector<std::array<double, 2>>> & img_pts,
  const std::vector<gtsam::Pose3> cam_world_poses,
  const Intrinsics & intrinsics, DistortionModel distortion_model,
  const DistortionCoefficients & distortion_coefficients,
  const std::string & fname)
{
  std::ofstream proj_file;
  const auto all_pts = computeAllProjectedPoints(
    world_pts, cam_world_poses, intrinsics, distortion_model,
    distortion_coefficients);
  if (!fname.empty()) {
    proj_file.open(fname, std::ios_base::app);
  }

  double sum_err{0};
  double max_err{-1e10};
  size_t max_idx{0};
  for (size_t i = 0; i < all_pts.size(); i++) {
    const auto & ic = all_pts[i];
    for (size_t k = 0; k < ic.size(); k++) {
      const std::array<double, 2> res{
        {ic[k][0] - img_pts[i][k][0], ic[k][1] - img_pts[i][k][1]}};
      if (!fname.empty()) {
        proj_file << std::fixed << std::setprecision(3) << cam_idx << " "
                  << times[i] << " " << ic[k][0] << " " << ic[k][1] << " "
                  << img_pts[i][k][0] << " " << img_pts[i][k][1] << std::endl;
      }
      const double err = res[0] * res[0] + res[1] * res[1];
      sum_err += err;
      if (err > max_err) {
        max_err = err;
        max_idx = i * world_pts.size() + k;
      }
    }
  }
  return {sum_err, max_err, max_idx};
}
}  // namespace diagnostics
}  // namespace multicam_imu_calib

File: multicam_imu_calib/src/diagnostics.py
#!/usr/bin/env python3
# -----------------------------------------------------------------------------
# Copyright 2024 Bernd Pfrommer <bernd.pfrommer@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
#

import argparse

import matplotlib.pyplot as plt
import numpy as np


def main(fname):
    a = np.loadtxt(fname)
    cam_idx = np.unique(a[:, 0]).astype(np.uint32)
    cutoff = 5
    for i in cam_idx:
        pts = a[a[:, 0].astype(np.uint32) == i, 2:]
        dx = pts[:, 0] - pts[:, 2]
        dy = pts[:, 1] - pts[:, 3]
        hx, bx = np.histogram(dx[np.abs(dx) < np.std(dx) * cutoff], bins=100, density=True)
        plt.plot(0.5 * (bx[:-1] + bx[1:]), hx, label='error in x')
        hy, by = np.histogram(dy[np.abs(dy) < np.std(dy) * cutoff], bins=100, density=True)
        plt.plot(0.5 * (by[:-1] + by[1:]), hy, label='error in y')
        plt.legend()
        plt.title(f'error distribution for camera {i}')
        plt.show()

        plt.scatter(pts[:, 0], pts[:, 1], s=8, label='image points')
        plt.scatter(pts[:, 2], pts[:, 3], s=8, label='projected points')
        plt.title(f'projection error cam {i}')
        plt.show()


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='calibration diagnostics.')
    parser.add_argument(
        '--projection_file_name',
        '-f',
        help='name of file with projected points',
        required=True,
        default=None,
    )
    args = parser.parse_args()
    main(args.projection_file_name)
